fix(compareArray): fix adding duplicate items to arrays not generating add operation

epelc opened the pull request:
If you added duplicate items to an array `compareArray` would not generate an add operation because it previously just checked that the object was in the array. Objects are now marked(via a map) that they have already been used and should be skipped. This fixes `compareArray` so that it can detect and generate proper patches when adding duplicate objects.

See `jsonpatch_array_test.go` for an example of when this happens.

```
PASS
coverage: 77.0% of statements
ok github.com/mattbaird/jsonpatch	0.042s
```

epelc committed:
fix(compareArray): fix adding duplicate items to arrays not generating a add operation

Previously it was only checked if an object was in an array. We now keep track of indexes of objects that have been found and only allow an object to be found once. This fixes adding duplicate arrays because the objects are essentially seen as being different now.

epelc commented:
Please wait to merge. I need to also fix this issue for removing elements. I didn't release there was the same issue.

epelc committed:
fix(compareArray): keep track of already found indexes so we can handle removing duplicates

epelc committed:
Improve compareArray

epelc commented:
This is ready to merge now.

NOTE I've added an extra test for modifying arrays with duplicate objects. In one case it will remove objects at the wrong index(if they match the object to be removed). See [here](https://github.com/mattbaird/jsonpatch/pull/4/commits/19d66f0e2e24a56249d7399b45eb8f5651d8e2ea#diff-9692ce33b581ed479856ed3b7b1475b5R57). This bug also existed before any of my changes.

It is a little complicated to fix because you have to switch from using `DeepEqual` to direct index comparisons but that leads to usually shifting items right to left when you remove something in the middle.

mattbaird commented:
hi @epelc any chance you could resolve the merge conflicts and I promise I'll merge quickly.

epelc committed:
Merge branch 'master' into fix/add-duplicate-object-to-array

# Conflicts:
#	jsonpatch.go

epelc commented:
Hi @mattbaird sorry for being unresponsive for 3 years. I think I missed the notification or something :)

Anyways I updated and rebased on the latest master as of now. All tests are working too.
